x-pack/apm-server/aggregation/txmetrics/aggregator.go

```diff
@@ -594,3 +649,103 @@ func transactionCount(tx *model.Transaction) float64 {
 	}
 	return 1
 }
+
+func writeLabels(w io.Writer, aggKey *transactionAggregationKey) {
+	if len(aggKey.labelKeys) == 0 && len(aggKey.numericLabels) == 0 {
+		return
+	}
+	for _, key := range aggKey.labelKeys {
+		label := aggKey.labels[key]
+		io.WriteString(w, key)
+		if label.Value != "" {
+			io.WriteString(w, label.Value)
+			continue
+		}
+		for _, v := range label.Values {
+			io.WriteString(w, v)
+		}
+	}
+	for _, key := range aggKey.numericLabelKeys {
+		label := aggKey.numericLabels[key]
+		io.WriteString(w, key)
+		if label.Value != 0 {
+			var b [8]byte
+			binary.LittleEndian.PutUint64(b[:], math.Float64bits(label.Value))
+			w.Write(b[:])
+			continue
+		}
+		for _, v := range label.Values {
+			var b [8]byte
+			binary.LittleEndian.PutUint64(b[:], math.Float64bits(v))
+			w.Write(b[:])
+		}
+	}
+}
+
+// equalLabels returns true if the labels are equal. The Global property is
+// ignored since only global labels are compared.
+func equalLabels(l, labels model.Labels) bool {
+	if len(l) != len(labels) {
+		return false
+	}
+	for key, localV := range l {
+		v, ok := labels[key]
+		if !ok {
+			return false
+		}
+		// If the slice value is set, ignore the Value field.
+		if len(v.Values) == 0 && v.Value != localV.Value {
+			return false
+		}
+		if len(v.Values) != len(localV.Values) {
+			return false
+		}
+		for _, value := range v.Values {
+			var found bool
+			for _, localValue := range localV.Values {
+				if value == localValue {
+					found = true
+					break
+				}
+			}
+			if !found {
+				return false
+			}
+		}
```

Comment:
```suggestion
		for i, value := range v.Values {
			if localV.Values[i] != value {
				return false
			}
		}
```

Do we need to treat array values as sets for comparison? I think we should consider them to be equal if they are exactly the same order.

Reply:
In that case slices that contain the exact same items but are out of order won't be equal:

```go
{"a": {Values: []string{"b", "a"}}}
```
```go
{"a": {Values: []string{"a", "b"}}}
```

Is that what we want? That seems counter intuitive to me.

Reply:
I haven't seen this explicitly specified anywhere, but as far as I've seen OTel will always maintain ordering of array-type attribute values. I don't think we should start treating them as sets, as this will lead to surprises if some OTel instrumentation is relying on the ordering being maintained. In practice I doubt it will matter.

The other reason is to make the code simpler and faster :)

Reply:
I see, thanks for the extra information. I'll change the code se we handle the order as intended.